compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -300,60 +302,84 @@ fn negative_impl<'cx, 'tcx>(
-
     // Create an infcx, taking the predicates of impl1 as assumptions:
     tcx.infer_ctxt().enter(|infcx| {
-        // Normalize the trait reference. The WF rules ought to ensure
-        // that this always succeeds.
-        let impl1_trait_ref = match traits::fully_normalize(
-            &infcx,
-            FulfillmentContext::new(),
-            ObligationCause::dummy(),
-            impl1_env,
-            impl1_trait_ref,
-        ) {
-            Ok(impl1_trait_ref) => impl1_trait_ref,
-            Err(err) => {
-                bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
-            }
-        };
+        // create a parameter environment corresponding to a (placeholder) instantiation of impl1
+        let impl1_env = tcx.param_env(impl1_def_id);
+
+        if let Some(impl1_trait_ref) = tcx.impl_trait_ref(impl1_def_id) {
+            // Normalize the trait reference. The WF rules ought to ensure
+            // that this always succeeds.
+            let impl1_trait_ref = match traits::fully_normalize(
+                &infcx,
+                FulfillmentContext::new(),
+                ObligationCause::dummy(),
+                impl1_env,
+                impl1_trait_ref,
+            ) {
+                Ok(impl1_trait_ref) => impl1_trait_ref,
+                Err(err) => {
+                    bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
+                }
+            };
+
+            // Attempt to prove that impl2 applies, given all of the above.
+            let selcx = &mut SelectionContext::new(&infcx);
+            let impl2_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl2_def_id);
+            let (impl2_trait_ref, obligations) =
+                impl_trait_ref_and_oblig(selcx, impl1_env, impl2_def_id, impl2_substs);
+
+            !obligations_satisfiable(
+                &infcx,
+                impl1_env,
+                impl1_def_id,
+                impl1_trait_ref,
+                impl2_trait_ref,
+                obligations,
+            )
+        } else {
+            let ty1 = tcx.type_of(impl1_def_id);
+            let ty2 = tcx.type_of(impl2_def_id);
+
+            !obligations_satisfiable(&infcx, impl1_env, impl1_def_id, ty1, ty2, iter::empty())
+        }
+    })
+}
 
-        // Attempt to prove that impl2 applies, given all of the above.
-        let selcx = &mut SelectionContext::new(&infcx);
-        let impl2_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl2_def_id);
-        let (impl2_trait_ref, obligations) =
-            impl_trait_ref_and_oblig(selcx, impl1_env, impl2_def_id, impl2_substs);
-
-        // do the impls unify? If not, not disjoint.
-        let Ok(InferOk { obligations: more_obligations, .. }) = infcx
-            .at(&ObligationCause::dummy(), impl1_env)
-            .eq(impl1_trait_ref, impl2_trait_ref)
-        else {
+fn obligations_satisfiable<'cx, 'tcx, T: Debug + ToTrace<'tcx>>(
```

Comment:
I kind of dislike this abstraction, the other thing I could do is kind of the opposite, have a function returning 2 `T: ToTrace` and obligations from traits or empty otherwise and then rollup all obligations_satisfiable code inside `negative_impl`.